File: bms/go/geo/geo_r2_Rect_Union/geo_r2_Rect_Union.go
package test

import (
	"math"
)

var ()

const ()

func (r geo_r2_Rect) Union(other geo_r2_Rect) geo_r2_Rect {
	return geo_r2_Rect{r.X.Union(other.X), r.Y.Union(other.Y)}
}

func (i geo_r1_Interval) Union(other geo_r1_Interval) geo_r1_Interval {
	if i.IsEmpty() {
		return other
	}
	if other.IsEmpty() {
		return i
	}
	return geo_r1_Interval{math.Min(i.Lo, other.Lo), math.Max(i.Hi, other.Hi)}
}

func (i geo_r1_Interval) IsEmpty() bool { return i.Lo > i.Hi }

type geo_r2_Rect struct {
	X, Y geo_r1_Interval
}

type geo_r1_Interval struct {
	Lo, Hi float64
}

File: translations_only/success/go/geo/geo_r2_Rect_Union/gpt4/BaseRepair/geo_r2_Rect_Union.rs

use std::f64;

struct GeoR2Rect {
    x: GeoR1Interval,
    y: GeoR1Interval,
}

struct GeoR1Interval {
    lo: f64,
    hi: f64,
}

impl GeoR2Rect {
    fn union(&self, other: &GeoR2Rect) -> GeoR2Rect {
        GeoR2Rect {
            x: union_r1(&self.x, &other.x),
            y: union_r1(&self.y, &other.y),
        }
    }
}

impl GeoR1Interval {
    fn is_empty(&self) -> bool {
        self.lo > self.hi
    }
}

fn union_r1(a: &GeoR1Interval, b: &GeoR1Interval) -> GeoR1Interval {
    if a.is_empty() {
        return GeoR1Interval { lo: b.lo, hi: b.hi };
    }
    if b.is_empty() {
        return GeoR1Interval { lo: a.lo, hi: a.hi };
    }
    GeoR1Interval {
        lo: a.lo.min(b.lo),
        hi: a.hi.max(b.hi),
    }
}
